buildtools/maven/maven.go

```diff
@@ -59,27 +60,75 @@ func (m *Maven) Compile(dir string) error {
 	return err
 }
 
-func Modules(dir string) ([]string, error) {
-	var pom Manifest
-	err := files.ReadXML(&pom, filepath.Join(dir, "pom.xml"))
+// A MvnModule can identify a Maven project with a groupId:artifactId string, a nested
+type MvnModule struct {
+	// Name is taken from the module's POM file.
+	Name string
+
+	// Target is the relative path from the root of the FOSSA project to the module's manifest file or to
+	// the directory containing the module's "pom.xml". The target may name a file other than "pom.xml" and
+	// it may even be the groupId:artifactId string identifying the "Maven project".
+	Target string
+
+	// Path is the relative path from the root of the FOSSA project to the module.
+	Dir string
+
+	GroupId    string
+	ArtifactId string
+}
+
+// Modules returns a list specifying the Maven module at path, which may name a file or directory, and all the
+// Maven modules nested below it. The Target field of each MvnModule is set to the directory or manifest file
+// describing the module. The visited manifest files are listed in the checked map.
+func Modules(path string, checked map[string]bool) ([]MvnModule, error) {
+	cwd, err := os.Getwd()
 	if err != nil {
-		return nil, errors.Wrap(err, "could not read POM file")
+		return nil, errors.Wrap(err, "could not get CWD")
 	}
 
-	var modules []string
-	for _, module := range pom.Modules {
-		children, err := Modules(filepath.Join(dir, module))
+	dir := path
+	pomFile := path
+	if strings.HasSuffix(path, ".xml") {
```

Comment:
When does this happen?

Reply:
Two ways:
1. A `module` can be specified directly in the `.fossa.yml` to have the `target: my-project.xml` or whatever.
2. In the `<modules>` of a POM file you can have paths to an actual such manifest file, not necessarily a directory.

My comment at the top of PR thread discusses this.

Reply:
Oh okay -- didn't know that <module> could point to an xml file. Rather than using `filepath.Rel`, then, can we use `filepath.Dir`? prevents us from having to pass in the base directory altogether 🙂

Reply:
sorry I'm not following. We still want `path` in `.fossa.yml` to be the relative directory of the module

Reply:
I think `path` here is already relative to the basedir -- so this can be changed to `dir = filepath.Dir(path)`. See the `filepath.Join` on 125, which applies to both modules pointing to directories and xml files

(If that's not the case, then we'd need a `filepath.Rel(cwd, ...)` in the `else` branch as well)

Reply:
I see what you're saying. We both understand that here `dir` needs to be a path that is valid anywhere in the recursion from the fossa project root (`baseDir`). My code was just being very defensive (the `Dir` field in `MvnModule` is document as being the relative path from the root of the FOSSA project, so what if `Modules` is called with an absolute path for `dir`?)—but perhaps overly defensive as you point out.